Clean up rustdoc's `main()`

nnethercote opened the pull request:
It can be simplified and made more similar to rustc's `main()`.

r? @oli-obk

nnethercote committed:
Avoid an unnecessary thread creation at rustdoc startup.

rustdoc's `main()` immediately spawns a thread, M, with a large stack
(16MiB or 32MiB) on which it runs `main_args()`. `main_args()` does a
small amount of options processing and then calls
`setup_callbacks_and_run_in_default_thread_pool_with_globals()`, which
spawns it own thread, and M is not used further.

So, thread M seems unnecessary. However, it does serve a purpose: if the
options processing in `main_args()` panics, that panic is caught when M
is joined. So M can't simply be removed.

However, `main_options()`, which is called by `main_args()`, has a
`catch_fatal_errors()` call within it. We can move that call to `main()`
and change it to the very similar `catch_with_exit_code()`. With that in
place, M can be removed, and panics from options processing will still
be caught appropriately.

Even better, this makes rustdoc's `main()` match rustc's `main()`, which
also uses `catch_with_exit_code()`.

(Also note that the use of a 16MiB/32MiB stack was eliminated from rustc
in #55617.)

nnethercote committed:
Remove `setup_callbacks_and_run_in_default_thread_pool_with_globals()`.

It's a very thin wrapper around
`setup_callbacks_and_run_in_thread_pool_with_globals()` and it has a
single call site.

nnethercote commented:
@oli-obk: not sure if you're the best reviewer, I chose you because you did #55617 which is somewhat related. Feel free to forward the review to someone else if necessary.

oli-obk commented on src/librustdoc/lib.rs:
```diff
@@ -429,24 +421,29 @@ fn main_args(args: &[String]) -> i32 {
     let options = match config::Options::from_matches(&matches) {
         Ok(opts) => opts,
-        Err(code) => return code,
+        Err(code) => return if code == 0 { Ok(()) } else { Err(()) },
```
I think this should return `Err(ErrorReported)` so that the caller doesn't have to do the `== Ok(())` dance.

oli-obk commented on src/librustdoc/lib.rs:
```diff
@@ -429,24 +421,29 @@ fn main_args(args: &[String]) -> i32 {
                 diag.struct_err(&err).emit();
             }
-            1
+            Err(())
```
This could also bubble up `ErrorReported` from here.

oli-obk commented on src/librustdoc/lib.rs:
```diff
@@ -429,24 +421,29 @@ fn main_args(args: &[String]) -> i32 {
-        Ok(()) => 0,
+        Ok(()) => Ok(()),
         Err(err) => {
             if !err.is_empty() {
```
Pre-existing... this feels very weird, either an error was reported, then we should be bubbling up an `ErrorReported`, or not, then we emit something, but I don't know what `is_empty` means here.

nnethercote replied:
Using `ErrorReported` is the proper way to do this. I have added a commit that does so.

nnethercote committed:
Be smarter about error handling in `run()`.

`run()` returns `Result<(), String>`. But on failure it always returns
an empty string, and then `wrap_return()` treats an empty string
specially, by not reporting the error.

It turns out we already have the `ErrorReported` type for this sort of
behaviour. This commit changes `run()` to use it.

nnethercote commented:
I have addressed all the comments.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 5f8a11279d75dc0078edbd8fa9d1fd1184c3f4ec has been approved by `oli-obk`

<!-- @bors r=oli-obk 5f8a11279d75dc0078edbd8fa9d1fd1184c3f4ec -->
<!-- homu: {"type":"Approved","sha":"5f8a11279d75dc0078edbd8fa9d1fd1184c3f4ec","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit 5f8a11279d75dc0078edbd8fa9d1fd1184c3f4ec with merge 1d69e3b1d753951bc7df0f02d6fd4719065d98c3...
<!-- homu: {"type":"BuildStarted","head_sha":"5f8a11279d75dc0078edbd8fa9d1fd1184c3f4ec","merge_sha":"1d69e3b1d753951bc7df0f02d6fd4719065d98c3"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/948819746), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35805)
Approved by: oli-obk
Pushing 1d69e3b1d753951bc7df0f02d6fd4719065d98c3 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35805","checks-actions":"https://github.com/rust-lang-ci/rust/runs/948819746"},"merge_sha":"1d69e3b1d753951bc7df0f02d6fd4719065d98c3"} -->
